ethexe/blob-loader/src/local.rs

```diff
@@ -0,0 +1,163 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2025 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{BlobData, BlobLoaderEvent, BlobLoaderService};
+use anyhow::{anyhow, Result};
+use ethexe_common::db::{CodesStorage, OnChainStorage};
+use ethexe_db::Database;
+use futures::{future::BoxFuture, stream::FusedStream, FutureExt, Stream};
+use gprimitives::CodeId;
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    sync::Arc,
+    task::Poll,
+};
+use tokio::sync::RwLock;
+
+#[derive(Clone)]
+pub struct LocalBlobStorage {
+    inner: Arc<RwLock<HashMap<CodeId, Vec<u8>>>>,
+    db: Database,
+}
+
+impl LocalBlobStorage {
+    pub fn new(db: Database) -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(HashMap::new())),
+            db,
+        }
+    }
+    pub async fn add_code(&self, code_id: CodeId, code: Vec<u8>) {
+        let mut storage = self.inner.write().await;
+        if storage.contains_key(&code_id) {
+            return;
+        }
+
+        storage.insert(code_id, code);
+    }
+
+    pub async fn get_code(self, code_id: CodeId) -> Result<BlobData> {
+        let storage = self.inner.read().await;
+
+        let code = match storage.get(&code_id) {
+            Some(code) => code.clone(),
+            None => self.db.original_code(code_id).ok_or({
+                log::error!("local storage: {storage:?}");
+                anyhow!("expect code for {code_id} exists in db")
+            })?,
```

Comment:
Do we really need this? Looks like access for code in db is not supposed to be here